### PR TITLE
IO: add `IO/Compression` directory

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #8691

Problem Summary:

### What is changed and how it works?

1. move compression related code under `IO/Compression/`
2. move `BaseFile/` to `IO/`
3. remove some useless files

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/IO/CompactContext.cpp [hunk 1]
```diff
@@ -38,8 +38,8 @@ extern const int UNKNOWN_FORMAT;
 extern const int FORMAT_VERSION_TOO_OLD;
 } // namespace ErrorCodes
 
-CompactWriteCtx::CompactWriteCtx(std::string compact_path_, size_t buffer_size)
-    : file_name(compact_path_)
+CompactWriteCtx::CompactWriteCtx(std::string part_path, size_t buffer_size)
+    : file_name(part_path)
```
Comment (CalvinNeo) [inline dbms/src/IO/CompactContext.cpp:42]: Why not use `std::move(part_path)` here?

File: dbms/src/IO/CompactContext.cpp [hunk 2]
```diff
@@ -64,16 +64,16 @@ size_t CompactWriteCtx::flushAllMarks()
 
 void CompactWriteCtx::writeFooter(size_t rows_count, size_t offset_of_all_marks)
 {
-    size_t footerOffset = plain_hashing->count();
+    size_t footer_offset = plain_hashing->count();
     writeIntBinary(mark_map.size(), *plain_hashing);
-    for (auto it = mark_map.begin(); it != mark_map.end(); it++)
+    for (auto & it : mark_map)
```
Comment (CalvinNeo) [inline dbms/src/IO/CompactContext.cpp:69]: Is it OK to use `const & it`? Seems the `it` is not mutated.

File: dbms/src/IO/CompactContext.cpp [hunk 3]
```diff
@@ -119,8 +119,9 @@ void CompactWriteCtx::writeEOF(std::shared_ptr<HashingWriteBuffer> hashing)
 }
 
 CompactReadCtx::CompactReadCtx(std::string compact_path_)
+    : compact_path(compact_path_)
```
Comment (CalvinNeo) [inline dbms/src/IO/CompactContext.cpp:122]: ditto

File: dbms/src/IO/Compression/ZlibInflatingReadBuffer.cpp [hunk 4]
```diff
@@ -25,14 +25,15 @@ ZlibInflatingReadBuffer::ZlibInflatingReadBuffer(
     size_t alignment)
     : BufferWithOwnMemory<ReadBuffer>(buf_size, existing_memory, alignment)
     , in(in_)
+    , zstr()
```
Comment (CalvinNeo) [inline dbms/src/IO/Compression/ZlibInflatingReadBuffer.cpp:28]: If we have `zstr()` here, maybe we can remove the lines from  `next_in` to `avail_out`?

Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/ZlibInflatingReadBuffer.cpp]: Actually those files are useless now, removed

File: dbms/src/IO/CompactContext.cpp (hunk 1 above)
Comment (CalvinNeo) [reply dbms/src/IO/CompactContext.cpp]: Is this also deleted?

Comment (Lloyd-Pottiger) [reply dbms/src/IO/CompactContext.cpp]: yes

**Review by CalvinNeo — APPROVED** [review]
lgtm

File: dbms/src/Common/tests/arena_with_free_lists.cpp
Comment (JaySon-Huang) [inline dbms/src/Common/tests/arena_with_free_lists.cpp:1]: nit: also remove other legacy "test" programs and its files that is not run actually

Comment (Lloyd-Pottiger) [reply dbms/src/Common/tests/arena_with_free_lists.cpp:1]: ok, in next PR.

**Pull request merged**